File: elixir/list-ops/lib/list_ops.ex
defmodule ListOps do
  def count([]), do: 0
  def count([_ | tail]), do: 1 + count(tail)

  def reverse(list), do: reverse(list, [])

  def reverse([], reversed), do: reversed
  def reverse([head | tail], reversed), do: reverse(tail, [head | reversed])

  def map([], _), do: []
  def map([head | tail], func), do: [func.(head) | map(tail, func)]

  def filter([], _), do: []

  def filter([head | tail], pred) do
    case pred.(head) do
      true ->
        [head | filter(tail, pred)]

      false ->
        filter(tail, pred)
    end
  end

  def reduce([], acc, _), do: acc

  def reduce([head | tail], acc, func) do
    reduce(tail, func.(head, acc), func)
  end

  def append(list_a, list_b), do: list_a ++ list_b

  def concat([]), do: []
  def concat([head | tail]), do: head ++ concat(tail)
end

File: ruby/list-ops/list_ops.rb
class ListOps
  def self.arrays(list)
    len = 0
    for _ in list
      len += 1
    end
    len
  end

  def self.reverser(list)
    reversed = []
    len = arrays(list) - 1
    for i in 0..len
      reversed << list[len - i]
    end
    reversed
  end

  def self.concatter(list_a, list_b)
    [*list_a, *list_b]
  end

  def self.mapper(list)
    result = []
    for item in list
      result << yield(item)
    end
    result
  end

  def self.filterer(list)
    result = []
    for item in list
      result << item if yield(item)
    end
    result
  end

  def self.reduce(list, initial)
    acc = initial
    for item in list
      acc = yield(acc, item)
    end
    acc
  end

  def self.sum_reducer(nums)
    reduce(nums, 0) { |acc, v| acc + v }
  end

  def self.factorial_reducer(nums)
    reduce(nums, 1) { |acc, v| acc * v }
  end
end
